File: Editor/lib/components/component_factory.dart
import 'package:editor/components/components.dart';
import 'package:editor/components/script.dart';
import 'package:editor/components/transform.dart';

typedef CreationFunctionType = Function(Object);

enum ComponentType {
  trasnsform,
  script,
}

extension ComponentFactory on Component {
  static final List<CreationFunctionType> _functions = [
    (data) => Transform(),
    (data) => Script(name: data as String)
  ];

  static CreationFunctionType getCreationFunction(ComponentType componentType) {
    return _functions[componentType.index];
  }

  ComponentType toEnumType() {
    if (this is Transform) {
      return ComponentType.trasnsform;
    } else if (this is Script) {
      return ComponentType.script;
    } else {
      throw ArgumentError('Unknown component type');
    }
  }
}

File: Editor/Components/ComponentFactory.cs
using Editor.Common.Enums;
using System.Diagnostics;

namespace Editor.Components
{
    static class ComponentFactory
    {
        private static readonly Func<Entity, object, Component>[] _functions =
        [
            (entity, data) => new Transform(entity),
            (entity, data) => new Script(entity)
            {
                Name = (string)data,
            }
        ];

        public static Func<Entity, object, Component> GetCreationFunction(ComponentType componentType)
        {
            Debug.Assert((int)componentType < _functions.Length);

            return _functions[(int)componentType];
        }

        public static ComponentType ToEnumType(this Component c) =>
            c switch
            {
                Transform _ => ComponentType.TRANSFORM,
                Script _ => ComponentType.SCRIPT,
                _ => throw new ArgumentException("Unknown component type")
            };
    }
}
